### PR TITLE
signature: Move signature outside, use a layered approach for signature verification

### PR DESCRIPTION
**Pull request by harshavardhana**
Signature calculation has now moved out from being a package to
top-level as a layered mechanism.

In case of payload calculation with body, go-routines are initiated
to simultaneously write and calculate shasum. Errors are sent
over the writer so that the lower layer removes the temporary files
properly.

### CODE REVIEW

**Comment by harshavardhana** [comment]
Not ready for review yet!.

File: auth-handler.go [hunk 1]
```diff
@@ -97,12 +96,41 @@ func getRequestAuthType(r *http.Request) authType {
...
+	hash.Write(data)
+	return hash.Sum(nil)
+}
+
 // Verify if request has valid AWS Signature Version '4'.
-func isReqAuthenticated(sign *signature4.Sign, r *http.Request) (s3Error APIErrorCode) {
-	auth := sign.SetHTTPRequestToVerify(r)
+func isReqAuthenticated(r *http.Request) (s3Error APIErrorCode) {
+	if r == nil {
+		errorIf(probe.NewError(errInvalidArgument), "http request cannot be empty.", nil)
```
Comment (abperiasamy) [inline auth-handler.go]: HTTP request cannot be empty.

File: auth-handler.go [hunk 2]
```diff
@@ -97,12 +96,41 @@ func getRequestAuthType(r *http.Request) authType {
...
+		return ErrInternalError
+	}
+	payload, e := ioutil.ReadAll(r.Body)
+	if e != nil {
+		errorIf(probe.NewError(e), "Reading body failed.", nil)
```
Comment (abperiasamy) [inline auth-handler.go]: Unable to read HTTP body.

File: generic-handlers.go [hunk 3]
```diff
@@ -51,6 +46,11 @@ type redirectHandler struct {
 	locationPrefix string
 }
 
+// Private bucket.
+const (
+	privateBucket = "/minio"
+)
```
Comment (abperiasamy) [inline generic-handlers.go:52]: reservedBucket

**Comment by harshavardhana** [comment]
> reservedBucket

This change is there in another set of patch i didn't want to make it since it's not related with this one.

**Pull request merged**